datajoint_utilities/generic/typed.py

```diff
@@ -0,0 +1,652 @@
+    import_name: djt.StrNone = None,
+    must_exist: Literal[True] = True,
+) -> Path | NoReturn:
+    ...
+
+
+@overload
+def pkg_abspath(
+    import_name: djt.StrNone = None,
+    must_exist: Literal[False] = False,
+) -> Path | None:
+    ...
+
+
+def pkg_abspath(
+    import_name: djt.StrNone = None,
+    must_exist: Literal[True, False] = True,
+) -> Path | None | NoReturn:
+    import_name = import_name or pkg_name_from_frame(inspect.stack())
+    pkg_spec = find_spec(import_name)
+    if not pkg_spec:
+        if must_exist:
+            raise PackageNotFoundError(import_name)
+        return None
+    return Path(str(pkg_spec.origin)).parent
+
+
+def pkg_relpath(
+    file: djt.StrPath,
+    *,
+    package_import_name: djt.StrNone = None,
+    package_root_path: djt.StrPathNone = None,
+    append_to_package_root_path: djt.StrPathNone = None,
+) -> Path | None:
+    src_path = Path(file)
+    root_path = Path(package_root_path or pkg_abspath(package_import_name))
+    if append_to_package_root_path:
+        root_path /= append_to_package_root_path
+    src_is_rel = src_path.is_relative_to(root_path)
+    return src_path.relative_to(root_path) if src_is_rel else None
+
+
+@overload
+def as_path_parts(file_or_parts: djt.StrTuple) -> djt.StrTuple:
+    ...
+
+
+@overload
+def as_path_parts(file_or_parts: djt.StrPath) -> djt.StrTuple:
+    ...
+
+
+def as_path_parts(file_or_parts: djt.StrPathParts) -> djt.StrTuple:
+    """Like `Path().parts` but with no extension at end.
+
+    Args:
+        file_or_parts (djt.StrPathParts): A sequence of path parts or a path string to get parts from.
+
+    Raises:
+        TypeError: Incorrect type of argument.
+
+    Returns:
+        djt.StrTuple: A sequence of path parts.
+    """
+    if file_or_parts is None:
+        return djt.StrTuple()
+
+    if isinstance(file_or_parts, (tuple, list)):
+        parts = file_or_parts
+
+    elif djt.is_pathstr(file_or_parts):
+        file_or_parts = Path(file_or_parts)
+        parts = (*file_or_parts.parts[:-1], file_or_parts.stem)
+    else:
+        raise TypeError(
+            f"argument must be of type <{djt.StrPathParts}>, "
+            f"not '{type(file_or_parts)}'"
+        )
+
+    return (*filter(None, parts),)
+
+
+def chain_path_parts(*args: djt.StrPathParts, sep: str = ".") -> str:
+    part_gen = (as_path_parts(parts) for parts in args)
+    return sep.join(itertools.chain.from_iterable(part_gen))
+
+
+def filepath_from_frame(source: str | Path | djt.FrameStack | None) -> Path | None:
+    if djt.is_frame_stack(source):
+        return calling_frame_file(source)
+    if djt.is_pathstr(source) and source:
+        return Path(source)
+    return None
+
+
+@dataclass
+class SourceFileInfo:
+    file: Path = field(default_factory=Path)
+    pkg_name: str = ""
+    pkg_relpath: Path = field(default_factory=Path)
+
+
+# datetime functions -------------------------------------------------------------------
+def datetime_to_utcaware(
```

Comment:
What's UTC aware?  The `datetime` package must provide a standard way for converting into UTC.

Reply:
I couldn't find a solution that didn't involve importing the pytz package